snuba/query/expressions.py

```diff
@@ -0,0 +1,134 @@
+from __future__ import annotations
+
+from abc import ABC
+from dataclasses import dataclass
+from typing import Callable, Iterable, Optional, Iterator
+
+from snuba.query.nodes import AliasedNode
+from snuba.query.collections import NodeContainer
+
+
+class Expression(AliasedNode, ABC):
+    """
+    Abstract representation of a Query node that can be evaluated to a single value.
+    This can be a simple column, NULL or a nested expression, but not a condition.
```

Comment:
Technically according to many languages grammar, conditions are expression. I think this is a matter of naming, a condition cannot extend Expression as it is (since expresison can have an alias). I could add another level above the two but it does not seem useful since it would not be used anywhere.
(I am implying we are not supporting something like: SELECT a = b or c FROM sentry for now)

Reply:
I don't really understand why this distinction needs to get made at this level. The functionality that can be expressed (no pun intended) with an expression is a superset of the functionality of a condition, so why not just use the same type internally to represent them?

Reply:
The only difference right now is that Condition should not be an AliasedNode. but a FormattableNode, while an expression can be assigned an alias.
So Condition cannot just inherit from Expression.
In order to make Condition an Expression we would need :
```
class Expression:

class AliasedExpression(Expression, AliasedNode)

class Column(AliasedExpression)

class Condition(Expression)
```

This way the Expression class does not actually do anything. 
If the naming is confusing I am ok with dropping the term Expression and just having the AliasedExpression. What do you think ?

Reply:
This would still be a valid point if I turned AliasedNoe into a wrapper, since I would not want AliasedNode to be instantiated with a condition, which could become something like this when formatted:
```
SELECT ... WHERE (a = b) as c
```

Reply:
This should not be relevant anymore since now a Condition is also an Expression